From 22a91480408b959a5f5cd27817567a169429cf76 Mon Sep 17 00:00:00 2001
From: "Miguel A. Vico" <mvicomoya@nvidia.com>
Date: Wed, 25 Sep 2019 11:23:13 -0700
Subject: [PATCH] Fix display hotplugging/unplugging crashes (already on main
 stream)

Include below CL:
5c5f0272 compositor: Do not trigger invalid destructors when hotunplugging
620f68dc desktop-shell: Avoid NULL output dereference when getting surface label
---
 compositor/main.c     | 12 ++++++++++--
 desktop-shell/shell.c |  6 +++---
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/compositor/main.c b/compositor/main.c
index 144777d..800956a 100644
--- a/compositor/main.c
+++ b/compositor/main.c
@@ -1848,8 +1848,16 @@ wet_output_from_weston_output(struct weston_output *base)
 static void
 wet_output_destroy(struct wet_output *output)
 {
-	if (output->output)
-		weston_output_destroy(output->output);
+	if (output->output) {
+		/* output->output destruction may be deferred in some cases (see
+		 * drm_output_destroy()), so we need to forcibly trigger the
+		 * destruction callback now, or otherwise would later access
+		 * data that we are about to free
+		 */
+		struct weston_output *save = output->output;
+		wet_output_handle_destroy(&output->output_destroy_listener, save);
+		weston_output_destroy(save);
+	}
 
 	wl_list_remove(&output->link);
 	free(output);
diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index 93b1c70..7b55af2 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -521,7 +521,7 @@ static int
 focus_surface_get_label(struct weston_surface *surface, char *buf, size_t len)
 {
 	return snprintf(buf, len, "focus highlight effect for output %s",
-			surface->output->name);
+			(surface->output ? surface->output->name : "NULL"));
 }
 
 /* no-op func for checking focus surface */
@@ -2985,7 +2985,7 @@ static int
 background_get_label(struct weston_surface *surface, char *buf, size_t len)
 {
 	return snprintf(buf, len, "background for output %s",
-			surface->output->name);
+			(surface->output ? surface->output->name : "NULL"));
 }
 
 static void
@@ -3065,7 +3065,7 @@ static int
 panel_get_label(struct weston_surface *surface, char *buf, size_t len)
 {
 	return snprintf(buf, len, "panel for output %s",
-			surface->output->name);
+			(surface->output ? surface->output->name : "NULL"));
 }
 
 static void
-- 
2.7.4

